src/lib/shim/src/tls.rs

```diff
@@ -309,105 +445,123 @@ pub unsafe fn unregister_current_thread() {
+/// It would mean another map-lookup on every thread-local access, but it'd
+/// probably ok.
 struct ShimTlsVarStorage<T> {
     // Whether the var has been initialized for this thread.
     //
-    // This is effectively a bool with C-style semantics.  We use an i8 because
-    // we need the "0 value" of this struct to indicate that it's not
-    // initialized. `bool` is documented as *casting* to 0 or 1, but it's
-    // possible (if unlikely) that the underlying representation in memory isn't
-    // actually 0 and 1.
-    //
-    // Likewise we can't use an `Option<T>` for `value`, since we can't guarantee
-    // that the "0 value" will be a valid representation of `None`.
-    //
-    // We need `UnsafeCell` for both so that we can initialize, despite this being
-    // stored in a larger array over which we don't have exclusive access.
-    initd: UnsafeCell<i8>,
+    // For `bool`, the bit pattern 0 is guaranteed to represent
+    // `false`, and `Cell` has the same layout as its inner type. Hence,
+    // interpreting 0-bytes as `Self` is sound and correctly indicates that it hasn't
+    // been initialized.
+    // <https://doc.rust-lang.org/std/cell/struct.Cell.html#memory-layout>
+    // <https://doc.rust-lang.org/reference/types/boolean.html>
+    initd: Cell<bool>,
     value: UnsafeCell<MaybeUninit<T>>,
 }
 
-// An initializer for internal use with `LazyLock`. We need an explicit type
-// instead of just a closure so that we can name the type  in `ShimTlsVar`'s
-// definition.
-struct OffsetInitializer<T, F> {
-    _phantom_t: PhantomData<T>,
-    _phantom_f: PhantomData<F>,
+/// An initializer for internal use with `LazyLock`. We need an explicit type
+/// instead of just a closure so that we can name the type  in `ShimTlsVar`'s
+/// definition.
+struct OffsetInitializer {
+    align: usize,
+    size: usize,
 }
 
-impl<T, F> OffsetInitializer<T, F> {
-    pub const fn new() -> Self {
+impl OffsetInitializer {
+    pub const fn new<T>() -> Self {
         Self {
-            _phantom_t: PhantomData,
-            _phantom_f: PhantomData,
+            // FIXME: should be of storage.
+            align: core::mem::align_of::<ShimTlsVarStorage<T>>(),
+            size: core::mem::size_of::<ShimTlsVarStorage<T>>(),
         }
     }
 }
 
-impl<T, F> lazy_lock::Producer<usize> for OffsetInitializer<T, F>
-where
-    F: Fn() -> T,
-{
+impl lazy_lock::Producer<usize> for OffsetInitializer {
     // Finds and assigns the next free and suitably aligned offset within
     // thread-local-storage for a value of type `T`, initialized with function
     // `F`.
     fn initialize(self) -> usize {
+        // The alignment we ensure here is an offset from the base of a [`ShimThreadLocalStorage`].
+        // It won't be meaningful if [`ShimThreadLocalStorage`] has a smaller alignment requirement
+        // than this variable.
+        assert!(self.align <= core::mem::align_of::<ShimThreadLocalStorage>());
```

Comment:
I still don't understand this part. What happens if the alignment were larger than `ShimThreadLocalStorage`?